File: js/chat/src/EventEmitter.ts
/**
 * Tiny typed event emitter — framework-agnostic base for NoLagChat and ChatRoom.
 *
 * EventMap is a record of event name → tuple of handler arguments.
 * e.g. { message: [ChatMessage]; typing: [{ users: ChatUser[] }] }
 */
export class EventEmitter<EventMap extends { [K in keyof EventMap]: unknown[] }> {
  private _handlers = new Map<keyof EventMap, Set<(...args: any[]) => void>>();

  /**
   * Register an event handler.
   */
  on<K extends keyof EventMap>(event: K, handler: (...args: EventMap[K]) => void): this {
    if (!this._handlers.has(event)) {
      this._handlers.set(event, new Set());
    }
    this._handlers.get(event)!.add(handler);
    return this;
  }

  /**
   * Remove a specific handler, or all handlers for an event.
   */
  off<K extends keyof EventMap>(event: K, handler?: (...args: EventMap[K]) => void): this {
    if (handler) {
      this._handlers.get(event)?.delete(handler);
    } else {
      this._handlers.delete(event);
    }
    return this;
  }

  /**
   * Remove all handlers for all events.
   */
  removeAllListeners(): this {
    this._handlers.clear();
    return this;
  }

  /**
   * Emit an event to all registered handlers.
   */
  protected emit<K extends keyof EventMap>(event: K, ...args: EventMap[K]): void {
    const handlers = this._handlers.get(event);
    if (!handlers) return;
    for (const handler of handlers) {
      try {
        handler(...args);
      } catch (e) {
        console.error(`Error in ${String(event)} handler:`, e);
      }
    }
  }

  /**
   * Returns the number of handlers registered for an event.
   */
  listenerCount<K extends keyof EventMap>(event: K): number {
    return this._handlers.get(event)?.size ?? 0;
  }
}

File: python/agents/nolag_agents/event_emitter.py
from __future__ import annotations

from collections import defaultdict
from typing import Any, Callable
import traceback


class EventEmitter:
    """Typed event emitter matching the JS SDK pattern."""

    def __init__(self) -> None:
        self._listeners: dict[str, list[Callable[..., Any]]] = defaultdict(list)

    def on(self, event: str, handler: Callable[..., Any]) -> EventEmitter:
        if handler not in self._listeners[event]:
            self._listeners[event].append(handler)
        return self

    def off(self, event: str, handler: Callable[..., Any] | None = None) -> EventEmitter:
        if handler is None:
            self._listeners.pop(event, None)
        elif event in self._listeners:
            try:
                self._listeners[event].remove(handler)
            except ValueError:
                pass
        return self

    def remove_all_listeners(self) -> EventEmitter:
        self._listeners.clear()
        return self

    def _emit(self, event: str, *args: Any) -> None:
        for handler in list(self._listeners.get(event, [])):
            try:
                handler(*args)
            except Exception:
                traceback.print_exc()

    def listener_count(self, event: str) -> int:
        return len(self._listeners.get(event, []))
